src/Compilers/VisualBasic/Portable/Emit/SynthesizedPrivateImplementationDetailsSharedConstructor.vb

```diff
@@ -0,0 +1,96 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports System.Collections.Generic
+Imports System.Collections.Immutable
+Imports System.Diagnostics
+Imports System.Linq
+Imports Microsoft.CodeAnalysis.CodeGen
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+    Friend NotInheritable Class SynthesizedPrivateImplementationDetailsSharedConstructor
+        Inherits SynthesizedConstructorSymbol
+
+        Private _containingModule As SourceModuleSymbol
+        Private _privateImplementationType As PrivateImplementationDetails
+
+        Friend Sub New(
+            containingModule As SourceModuleSymbol,
+            privateImplementationType As PrivateImplementationDetails,
+            diagnostics As DiagnosticBag,
+            voidType As NamedTypeSymbol
+        )
+            MyBase.New(Nothing, Nothing, True, False, Nothing, diagnostics)
+
+            _containingModule = containingModule
+            _privateImplementationType = privateImplementationType
+        End Sub
+
+        Friend Overrides Function GetBoundMethodBody(diagnostics As DiagnosticBag, Optional ByRef methodBodyBinder As Binder = Nothing) As BoundBlock
+            methodBodyBinder = Nothing
+
+            ' Initialize the payload root for for each kind of dynamic analysis instrumentation.
+            ' A payload root Is an array of arrays of per-method instrumentation payloads.
+            ' For each kind of instrumentation:
+            '
+            '     payloadRoot = New T(MaximumMethodDefIndex)() {}
+            '
+            ' where T Is the type of the payload at each instrumentation point, And MaximumMethodDefIndex Is the 
+            ' index portion of the greatest method definition token in the compilation. This guarantees that any
+            ' method can use the index portion of its own method definition token as an index into the payload array.
+
+            Dim payloadRootFields As IReadOnlyCollection(Of KeyValuePair(Of Integer, InstrumentationPayloadRootField)) = _privateImplementationType.GetInstrumentationPayloadRoots()
+            Debug.Assert(payloadRootFields.Count > 0)
+
+            Dim body As ArrayBuilder(Of BoundStatement) = ArrayBuilder(Of BoundStatement).GetInstance(2 + payloadRootFields.Count)
+
+            For Each payloadRoot As KeyValuePair(Of Integer, InstrumentationPayloadRootField) In payloadRootFields.OrderBy(Function(analysis) analysis.Key)
+
+                Dim analysisKind As Integer = payloadRoot.Key
+                Dim payloadArrayType As ArrayTypeSymbol = DirectCast(payloadRoot.Value.Type, ArrayTypeSymbol)
+
+                Dim maxMethodIndex As BoundExpression = New BoundMaximumMethodDefIndex(Syntax, factory.SpecialType(SpecialType.System_Int32))
```

Comment:
It looks like we could add TypeCompilationState as a parameter for this method. It is available in all places, but two. In those places we can simply create one in the same manner as it is done in MethodCompiler.VisitEmbeddedMethod and similarly assert

```
Debug.Assert(Not compilationState.HasSynthesizedMethods)
```

after we are done.

---
In reply to: [68326768](https://github.com/dotnet/roslyn/pull/12188#discussion_r68326768) [](ancestors = 68326768)